src/frontends/lean/elaborator.cpp

```diff
@@ -2884,350 +2869,405 @@ expr elaborator::instantiate_mvars(expr const & e, std::function<bool(expr const
-                    progress = true;
-                } else if (optional<name> default_value_fn = has_default_value(m_env, S_name, S_fname)) {
-                    expr fval = mk_field_default_value(m_env, full_S_fname, [&](name const & fname) {
-                        // just insert mvars for now, we will check for remaining ones in `reduce_and_check_deps` later
-                        if (auto mvar = field2mvar.find(fname)) {
-                            return some_expr(mk_as_is(instantiate_mvars(*mvar)));
-                        } else if (auto val = field2value.find(fname)) {
-                            return some_expr(mk_as_is(*val));
-                        } else {
-                            return none_expr();
-                        }
-                    });
-                    expr new_fval;
-                    expr new_fval_type;
-                    optional<expr> new_new_fval;
-                    std::tie(new_fval, new_fval_type, new_new_fval) = elaborate_arg(fval, expected_type, ref);
-                    if (new_new_fval) {
-                        /* delta- and beta-reduce `_default` definition */
-                        expr fval = *new_new_fval;
-                        buffer<expr> args;
-                        expr fn = get_app_args(fval, args);
-                        if (is_constant(fn)) {
-                            declaration decl = m_env.get(const_name(fn));
-                            expr default_val = instantiate_value_univ_params(decl, const_levels(fn));
-                            // clean up 'id' application inserted by `structure_cmd::declare_defaults`
-                            default_val = replace(default_val, [](expr const &e) {
-                                if (is_app_of(e, get_id_name(), 2)) {
-                                    return some_expr(app_arg(e));
-                                }
-                                return none_expr();
-                            });
-                            fval = mk_app(default_val, args);
-                            fval = head_beta_reduce(fval);
-                        }
 
-                        if (!reduce_and_check_deps(fval))
-                            return;
-                        new_new_fval = some_expr(fval);
+                    expr val_type = m_elab.infer_type(val);
+                    if (auto val2 = m_elab.ensure_has_type(val, val_type, expected_type, m_ref)) {
+                        /* Make sure mvar is assigned, even if val2 is a meta var as well.
+                         * This is important for termination and the `instantiate_mvars` call in `operator()`.
+                         * Note that `ensure_has_type` has already unified their types, so this should not result
+                         * in any missed unifications.
+                         */
+                        lean_always_assert(m_ctx.match(mvar, *val2));
```

Comment:
@leodemoura I ended up using `match` after all because, as noted, the existing `instantiate_mvars` call below already depends on the helper mvars being unified away.